### PR TITLE
Add mapParallel method to StreamMessage

### PR DESCRIPTION
**Pull request by sleeplesslord**
Motivation:
#3962 added `mapAsync`, which allows transforming a stream with an asynchronous operation. `mapAsync` will wait for futures to complete to guarantee that the order is maintained. For uses cases where the order doesn't matter it would be useful to offer another method that eagerly publishes futures as soon as they complete. Since there is less waiting overall there can be performance benefits to this approach.

Modifications:
- Modify `AsyncMapStreamMessage` to allow controlling how many concurrent requests it will allow
- Add `StreamMessage.mapParallel` to allow creating a `AsyncMapStreamMessage` from an existing `StreamMessage` using a given concurrency limit (or unlimited)
- Rewrite `StreamMessage.mapAsync` as the special case of `mapParallel` where `maxConcurrency == 1`

Result:
- You can now use `mapParallel` to modify a stream using an async operation with a configurable concurrency limit

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:585]: `mapAsyncParallel` looks long to me.
How about using `mapParallel` that can also explain the behavior of this API?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:585]: Should we give an option that limits max parallelism?
For example:
```java
// allow to concurrently execute up to 8
streamMessage.mapParallel(function, 8);
// unlimited
streamMessage.mapParallel(function);
```

Comment (sleeplesslord) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:585]: I think that's a good idea. In that case I can also rewrite `mapAsync` as `mapParallel(function, 1)`

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:585]: > I can also rewrite mapAsync as mapParallel(function, 1)

Good idea!

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:611]: `int` seems enough for concurrency.

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:606]: Could you update Javadoc?

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:588]:
```suggestion
        return mapParallel(function, Integer.MAX_VALUE);
```

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:566]:
```suggestion
        return mapParallel(function, 1);
```

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:105]:
```suggestion
        private int pendingFutures;
```

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:253]: Let's say we have the following condition:
```
n = 1
maxConcurrency = 10
requestedByDownstream, requestedFromUpstream, pendingFutures = 0
```
If so, `toRequest` becomes `10` and `requestedByDownstream` becomes `-10`.
Consequently, `upstream.request(10)` is called even if a `Subscriber` requests 1 element.

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:193]: Should we request more than 1 if `maxConcurrency` is available?

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (sleeplesslord) [reply core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:253]: Ah, somehow I forgot about the possibility that `n < maxConcurrency`

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (sleeplesslord) [reply core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:193]: If my thinking is correct we should only have 1 available slot here, since this code will be executed each time a future finishes. Assuming downstream has requested some large amount we will always have `maxConcurrency == requestedFromUpstream + pendingFutures`. Since `pendingFutures` decreases by 1 here we only have space to increase `requestedFromUpstream` by 1.

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:585]: Add `@UnstableApi`?

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:608]: Add `@UnstableApi`?

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:610]: nit: 
```suggestion
            int maxConcurrency) {
```

File: core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java:612]: Should we check that `maxConcurrency` is positive?

**Review by ikhoon — APPROVED** [review]
Nice work! Left only minor nits. 🚀

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:105]: nit; wondering if `requestedToUpstream` is more representative of what this variable is tracking

```suggestion
        private int requestedToUpstream;
```

**Review by jrhee17 — APPROVED** [review]
Looks really nice! Only left one nit question. Thanks @sleeplesslord ! 🙇 👍 🚀

File: core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java
Comment (sleeplesslord) [reply core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java:105]: I'm a little bit conflicted on this. I think both are valid (English is weird). I'll leave it up to you maintainers 😄

**Review by trustin — APPROVED** [review]
Thanks a lot for your high quality contribution, @sleeplesslord! 🙇

**Pull request merged**
